operators/src/processing/time_projection/mod.rs

```diff
@@ -0,0 +1,560 @@
+        .map_err(Into::into)
+    }
+
+    fn expand_feature_collection_result_inner(
+        feature_collection: FeatureCollection<G>,
+        step: TimeStep,
+        step_reference: TimeInstance,
+    ) -> Result<FeatureCollection<G>, TimeProjectionError> {
+        let time_intervals: Option<Result<Vec<TimeInterval>, TimeProjectionError>> =
+            feature_collection
+                .time_intervals()
+                .par_iter()
+                .with_min_len(128) // TODO: find good default
+                .map(|time_interval| expand_time_interval(step, step_reference, *time_interval))
+                // TODO: change to [`try_collect_into_vec`](https://github.com/rayon-rs/rayon/issues/713) if available
+                .try_fold_with(Vec::new(), |mut acc, time_interval| {
+                    acc.push(time_interval?);
+                    Ok(acc)
+                })
+                .try_reduce_with(|a, b| Ok([a, b].concat()));
+
+        match time_intervals {
+            Some(Ok(time_intervals)) => feature_collection
+                .replace_time(&time_intervals)
+                .context(error::CannotExpandFeatureTimeInterval),
+            Some(Err(error)) => Err(error),
+            None => Ok(feature_collection), // was empty
+        }
+    }
+}
+
+fn expand_query_rectangle(
+    step: TimeStep,
+    step_reference: TimeInstance,
+    query: VectorQueryRectangle,
+) -> Result<VectorQueryRectangle, TimeProjectionError> {
+    Ok(VectorQueryRectangle {
+        spatial_bounds: query.spatial_bounds,
+        time_interval: expand_time_interval(step, step_reference, query.time_interval)?,
+        spatial_resolution: query.spatial_resolution,
+    })
+}
+
+fn expand_time_interval(
+    step: TimeStep,
+    step_reference: TimeInstance,
+    time_interval: TimeInterval,
+) -> Result<TimeInterval, TimeProjectionError> {
+    let start = step.snap_relative_preserve_bounds(step_reference, time_interval.start());
+    let mut end = step.snap_relative_preserve_bounds(step_reference, time_interval.end());
+
+    // Since snap_relative snaps to the "left side", we have to add one step to the end
+    // This only applies if `time_interval.end()` is not the snap point itself.
+    if end < time_interval.end() {
+        end = match end + step {
+            Ok(end) => end,
+            Err(_) => TimeInstance::MAX, // `TimeInterval::MAX` can overflow
+        };
+    }
+
+    TimeInterval::new(start, end).context(error::CannotExpandQueryRectangle)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::{
+        engine::{MockExecutionContext, MockQueryContext},
+        mock::MockFeatureCollectionSource,
+    };
+    use chrono::NaiveDate;
+    use geoengine_datatypes::{
+        collections::MultiPointCollection,
+        primitives::{BoundingBox2D, MultiPoint, SpatialResolution, TimeGranularity, TimeInterval},
+        util::test::TestDefault,
+    };
+
+    #[test]
+    fn test_expand_query_time_interval() {
```

Comment:
please add a test where the snap point is larger then the data time.

Reply:
[1f5a87b](https://github.com/geo-engine/geoengine/pull/478/commits/1f5a87be9fe672115f5aedefc8d96583bfd7f8bf)